Automattic/newspack-ios
# Issues/91 progress ui

Refs #91 

This PR implements some UI improvements around syncing. 
- Asset cells now show a progress bar when their associated media file is uploading.
- Story and Asset lists now support pull to refresh.
- Creating/editing a story will attempt to sync to the server.
- Adding a new media asset will attempt to sync to the server.

Notes: 
- The sync buttons in the cells and story are probably going away in a future PR since pull to refresh will also handle any needed uploads.  I'm still thinking about this so I've left them for now.
- When adding/editing stories or assets there is a case where the sync coordinator might repeat a step. This is a quirk of the way I've implemented the queue and I'm content to leave it as the repeated step is basically a no op and handling the case would add complexity with out much gain.

@jleandroperez can I trouble you with another? 

Testing: 
Try this on a device with a test blog.  Due to a known issue with Metal the simulator probably won't work as well. 

Scenario 1: Creating Folders and Assets
- Create a few story folders. 
- Use the Photo feature and add a photo from your media library to the story.
- Confirm the Story and the Photo are uploaded.  Confirm you see a progress bar as the photo is uploaded.

Scenario 2: Pull to Refresh Folders
- Open the app to the stories list and keep the app open.
- In a browser, go to your test site's wp-admin and edit the title of the draft story. 
- In the app, pull to refresh the stories list. 
- Confirm that the story's title updates.

Scenario 3: Editing Folders
- In the app, rename a folder that has a remote draft.
- Once finished, check the draft story in a browser and confirm it's title has changed.

Scenario 4: Multiple media 
- In the app, view the folders list and add multiple photos to the current story from your device's media library.
- Quickly switch to the assets list for the current story and confirm you see the photos being uploaded and their progress indicators are displayed.
- Confirm multiple uploads complete successfully.

Scenario 5: Edit Story Asset
- Open the app to the assets list and keep the app open.
- In a browser, find the media object for an uploaded photo.
- Give the asset a caption.
- In the app, pull to refresh the assets list.
- Confirm the caption appears. 
- View the photo details and edit the caption again. 
- When finished, go back to the browser and confirm the edited caption appears.
